cvl/internal/yparser/yparser.go

```diff
@@ -695,110 +918,140 @@ func getErrorDetails() YParserError {
 	return errObj
 }
 
-func FindNode(root *YParserNode, xpath string) *YParserNode {
-	return (*YParserNode)(C.lyd_find_node((*C.struct_lyd_node)(root), C.CString(xpath)))
-}
-
 func GetModelNs(module *YParserModule) (ns, prefix string) {
 	return C.GoString(((*C.struct_lys_module)(module)).ns),
 		C.GoString(((*C.struct_lys_module)(module)).prefix)
 }
 
+// This function takes a when, must, or leafref path in its original form as
+// written in the YANG schema files, and converts it into its fully qualified
+// format.  The format in the YANG schema uses import prefixes, so we have to
+// replace each import prefix with the fully qualified module name.
+//
+// Libyang1 would do this for us automatically, but in Libyang3 we have to do
+// this conversion ourselves.
+//
+// In this implementation we are cheating a bit.  It would be quite a bit of
+// effort to tokenize everything, especially when and must clauses, so we rely
+// on the fact that the namespace can only contain alphanumeric or hypen
+// characters and will always begin with either "/" or "[" and end with ":".
+// So we extract each prefix using a regex, perform a lookup to determine the
+// module name then replace each instance of the prefix with the module name,
+// in a loop until all matched prefixes have been processed.
+//
+// Examples:
+//
+//	Leafref:
+//	  /po:sonic-portchannel/po:PORTCHANNEL/po:PORTCHANNEL_LIST/po:name
+//	     ->
+//	  /sonic-portchannel:sonic-portchannel/sonic-portchannel:PORTCHANNEL/sonic-portchannel:PORTCHANNEL_LIST/sonic-portchannel:name
+//
+//	Must:
+//	  (/cmn:operation/cmn:operation != 'CREATE') or
+//	  (count(/si:sonic-interface/si:INTERFACE/si:INTERFACE_IPADDR_LIST[si:ip_prefix=current()/../ip_prefix] [si:portname=(/svl:sonic-vlan/svl:VLAN_MEMBER/svl:VLAN_MEMBER_LIST[svl:name=current()]/svl:ifname)]) = 0)
+//	     ->
+//	  (/sonic-common:operation/sonic-common:operation != 'CREATE') or
+//	  (count(/sonic-interface:sonic-interface/sonic-interface:INTERFACE/sonic-interface:INTERFACE_IPADDR_LIST[sonic-interface:ip_prefix=current()/../ip_prefix] [sonic-interface:portname=(/sonic-vlan:sonic-vlan/sonic-vlan:VLAN_MEMBER/sonic-vlan:VLAN_MEMBER_LIST[sonic-vlan:name=current()]/sonic-vlan:ifname)]) = 0)
+func rewriteXPathPrefix(module *YParserModule, xpath string) string {
+	hasPrefix := make(map[string]bool)
+
+	re := regexp.MustCompile("[[/][A-Za-z0-9_-]+:")
```

Comment:
Same regex is compiled for every function call, which can be costly. Make it a global variable

Reply:
done